### PR TITLE
Add an exception to `Metrics/ParameterLists`.

### PR DESCRIPTION
**Pull request by reitermarkus**
Add an exception for `Metrics/ParameterLists` when passing a lambda/proc to `set_trace_func`.

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [ ] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Used the same coding conventions as the rest of the project.
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [ ] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [ ] All tests(`rake spec`) are passing.
* [x] The new code doesn't generate RuboCop offenses that are checked by `rake internal_investigation`.
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [ ] Updated cop documentation with `rake generate_cops_documentation` (required only when you've added a new cop or changed the configuration/documentation of an existing cop).

[1]: http://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: lib/rubocop/cop/metrics/parameter_lists.rb [hunk 1]
```diff
@@ -15,13 +15,21 @@ def on_args(node)
...
+
           add_offense(node) do
             self.max = count
           end
         end
 
         private
 
+        def argument_to_set_trace_func?(node)
+          node.parent && node.parent.parent &&
```
Comment (Drenmi) [inline lib/rubocop/cop/metrics/parameter_lists.rb]: You could probably use a node matcher, like:

```ruby
def_node_matcher :argument_to_set_trace_func?, <<-PATTERN
  ^^(send nil :set_trace_function _)
PATTERN
```

Each `^` ascends one level in the AST.

Comment (reitermarkus) [reply lib/rubocop/cop/metrics/parameter_lists.rb]: Thanks, I will try that.

**Comment by bbatsov** [comment]
It might be better if something like this is configurable. I hate hardcoding exceptions this way.

**Comment by bbatsov** [comment]
I've been thinking about this a moment more and I came up with a couple of related notes:

* This is a good opportunity to improve the pretty bad documentation of this cop
* Why is this flagging lambas and procs to begin with? Seems it should flag only instance and class method definitions. 

Don't know about the others, but I consider the current behavior a bug (that's why I added the note about documentation - right now it's hard to tell if this was intentional or not).

**Comment by Drenmi** [comment]
> It might be better if something like this is configurable. I hate hardcoding exceptions this way.

Makes sense. Perhaps an `IgnoreLambda` option?

> Why is this flagging lambas and procs to begin with? Seems it should flag only instance and class method definitions.

Good point. Could be an oversight, since both `def` and `lambda`/`proc` have a nested `args` node.

**Comment by reitermarkus** [comment]
Well, at least the name of the cop doesn't suggests that `lamba`/`proc` aren't covered by it, since they are still parameter lists. I wouldn't expect them to be excluded.

**Comment by bbatsov** [comment]
> Well, at least the name of the cop doesn't suggest that lamba/proc aren't covered by it, since they are still parameter lists. I wouldn't expect them to be excluded.

But I know we created this cop with the intent to avoid people defining methods with many parameters. It was never the intention to check call sites, which is what your patch is addressing. Flagging those seems pretty pointless as often they are part of the API of some other method.

**Comment by bbatsov** [comment]
> Good point. Could be an oversight, since both def and lambda/proc have a nested args node.

That's my assumption as well.

**Comment by bbatsov** [comment]
@reitermarkus Ping :-)

**Comment by reitermarkus** [comment]
I don't know what the plan is here now. Ignoring lambdas altogether?

**Comment by bbatsov** [comment]
Yeah, I'd just ignore lambdas completely.

**Comment by bbatsov** [comment]
The changes look OK. Just squash, rebase and add a changelog entry.

**Comment by bbatsov** [comment]
@reitermarkus Ping :-)

**Comment by reitermarkus** [comment]
Should be ready now.

File: CHANGELOG.md [hunk 2]
```diff
@@ -4,6 +4,7 @@
 
 ### New features
 
+* [#4848](https://github.com/bbatsov/rubocop/pull/4848): Exclude lambdas and procs from `Metrics/ParameterLists`. ([@reitermarkus][])
```
Comment (bbatsov) [inline CHANGELOG.md]: This should normally go under `Changes`.

**Comment by bbatsov** [comment]
@reitermarkus I think you forgot to add yourself to the end of the changelog.

**Comment by reitermarkus** [comment]
> I think you forgot to add yourself to the end of the changelog.

Ah! I couldn't figure out why my link wasn't working. 😄

**Comment by bbatsov** [comment]
👍

**Pull request merged**